File: backends/dotnet/src/StackverseBackend/Messages/MessageSeeder.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using StackverseBackend.Common;
using StackverseBackend.Data;

namespace StackverseBackend.Messages;

/// <summary>
/// SPEC rule 12: import the JSON seed files from `spec/messages` (language = filename),
/// inserting only `(key, language)` pairs that don't exist yet — runtime edits by
/// admins survive restarts. Seed inserts are not moderator actions, so they are
/// deliberately not audited.
/// </summary>
public static class MessageSeeder
{
    public static async Task SeedAsync(AppDbContext db, string messagesDir, ILogger logger)
    {
        var dir = Path.GetFullPath(messagesDir);
        if (!Directory.Exists(dir))
        {
            throw new InvalidOperationException(
                $"Message seed directory not found: {dir} — set SEED_MESSAGES_DIR to the spec/messages directory");
        }
        foreach (var file in Directory.EnumerateFiles(dir, "*.json").Order(StringComparer.Ordinal))
        {
            await SeedLanguageAsync(db, file, logger);
        }
    }

    private static async Task SeedLanguageAsync(AppDbContext db, string file, ILogger logger)
    {
        var language = Path.GetFileNameWithoutExtension(file);
        var entries = JsonSerializer.Deserialize<Dictionary<string, string>>(await File.ReadAllTextAsync(file))
            ?? throw new InvalidOperationException($"Message seed file is not a JSON object: {file}");
        var existing = await db.Messages.Where(m => m.Language == language).Select(m => m.Key).ToHashSetAsync();
        var now = Clock.UtcNow();
        var missing = entries.Where(entry => !existing.Contains(entry.Key))
            .Select(entry => new Message
            {
                Key = entry.Key,
                Language = language,
                Text = entry.Value,
                Description = null,
                CreatedAt = now,
                UpdatedAt = now,
            })
            .ToList();
        db.Messages.AddRange(missing);
        await db.SaveChangesAsync();
        logger.Event(LogLevel.Information, "message_seed_imported", "success",
            $"Message seed '{language}': {missing.Count} inserted, {entries.Count - missing.Count} already present",
            fields:
            [
                ("language", language),
                ("inserted", missing.Count),
                ("skipped", entries.Count - missing.Count),
            ]);
    }
}

File: backends/php-laravel/app/Services/MessageSeeder.php
<?php

namespace App\Services;

use App\Support\Logger;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;
use RuntimeException;

class MessageSeeder
{
    public function import(): void
    {
        $directory = (string) config('stackverse.seed_messages_dir');
        $files = glob(rtrim($directory, DIRECTORY_SEPARATOR).DIRECTORY_SEPARATOR.'*.json') ?: [];
        sort($files);
        if ($files === []) {
            throw new RuntimeException("Message seed directory not found: $directory");
        }

        foreach ($files as $file) {
            $language = pathinfo($file, PATHINFO_FILENAME);
            $entries = json_decode((string) file_get_contents($file), true, flags: JSON_THROW_ON_ERROR);
            if (! is_array($entries)) {
                throw new RuntimeException("Message seed file is not an object: $file");
            }

            $inserted = 0;
            foreach ($entries as $key => $text) {
                $changed = DB::affectingStatement(
                    'insert into messages (id, key, language, text, created_at, updated_at)
                     values (?, ?, ?, ?, clock_timestamp(), clock_timestamp())
                     on conflict (key, language) do nothing',
                    [(string) Str::uuid(), (string) $key, $language, (string) $text],
                );
                if ($changed) {
                    $inserted++;
                }
            }

            Logger::event(
                'info',
                'message_seed_imported',
                'success',
                "Message seed '$language': $inserted inserted, ".(count($entries) - $inserted).' already present',
                ['language' => $language, 'inserted' => $inserted, 'skipped' => count($entries) - $inserted],
            );
        }
    }
}
